pages/blog/add-reactivity-to-an-existing-source.mdx

```diff
@@ -0,0 +1,170 @@
+import withArticle from '../../ui/blog/article';
+
+export const meta = {
+  title: 'Add reactivity to an existing source',
+  author: 'jycouet',
+  tags: ['graphql', 'graphql-mesh', 'subscription'],
+  date: '2020-11-30',
+  description:
+    'How To extend an existing endpoint with subscriptions to bring reactivity to your application',
+  image: '/blog-assets/add-reactivity-to-an-existing-source/PimpMyGraphQL.png',
+};
+
+export default withArticle({ ...meta });
+
+# Introduction
+
+Working on the code generator for svelte plugin, I was looking for a public graphQL endpoint having queries, mutation and subscription for nice examples.
+Unfortunatly I didn't find anything out of the box, ready to use for my demos. So I decided to extend an existing source!
+
+# Starting point
+
+Across my researches I found one cool public endpoint regarding [SpaceX](https://api.spacex.land/graphql/). It has `Queries` and `Mutation` ready to use!
+So I decided to pimp this endpoint, and bring a bit of reactivity with subscriptions.
+
+The idea is to publish the new user when it's inserted with this `mutation`:
+
+```graphql
+mutation addUser {
+  insert_users(objects: { name: "jyc" }) {
+    affected_rows
+    returning {
+      name
+    }
+  }
+}
+```
+
+_How hard is to do this? Let's find out..._
+
+> Let's _Pimp this graphQL_!
+
+![Pimp my graphQL](/blog-assets/add-reactivity-to-an-existing-source/PimpMyGraphQL.gif)
+
+# Implementation
+
+<details>
+  <summary>Spoiler alert</summary>
+  Thanks to `graphql-mesh`, it's dead easy! 🚀🚀🚀
+</details>
+
+1. Create a new repo
+
+```bash
+mkdir newRepo
+cd newRepo
+yarn init
+```
+
+2. Add `graphql-mesh` packages
+
+```bash
+yarn add @graphql-mesh/cli @graphql-mesh/graphql graphql
+```
+
+3. Add `.meshrc.yaml` file in the root folder with
+
+```yaml
+sources:
+  - name: spacexGQL
+    handler:
+      graphql:
+        endpoint: https://api.spacex.land/graphql/
+
+additionalTypeDefs: |
+  extend type Mutation {
+    insert_users_and_publish(name: String!, rocket: String, twitter: String): users_mutation_response
+  }
+  extend type Subscription {
+    userAdded: users
+  }
+
+additionalResolvers:
+  - type: Subscription
+    field: userAdded
+    pubsubTopic: userAdded
+  - ./src/mutation-insert_users_and_publish.js
+```
+
+Where I
+
+- declare the graphQL source
+- extend `Mutation` to insert and publish the new user
+- extend `Subscription` to listen to user added users
+- add new resolver for `userAdded` listening to `userAdded` message
+- add new resolver for `insert_users_and_publish` in file `./src/mutation-insert_users_and_publish.js`
+
+4. Add `./src/mutation-insert_users_and_publish.js` file with all the logic for this new resolver.
+
+```js
+const resolvers = {
+  Mutation: {
+    insert_users_and_publish: async (
```

Comment:
Could this also be done without requiring a new field? Is it possible to do sth like an after/middleware for the existing delegated resolver?

Reply:
I was interested in this as well but didn't find it in `graphql-mesh`.
Now that you put this comment, I have to try another way that I didn't thought about before! I'll check now and give you feedback here sooooooon 👍

Reply:
Actually, it would be even much cooooooler!

But when I do:
```js
const resolvers = {
  Mutation: {
    insert_users: async (root, args, { spacexGQL, pubsub }) => {
      // use the existing mutation & overide fields to get for subscription
      const responseInsertUsers = await spacexGQL.api.insert_users(args, {
        fields: {
          affected_rows: true,
          returning: {
            id: true,
            name: true,
            rocket: true,
            timestamp: true,
            twitter: true,
          }
        },
      });

      // publish all users created at once
      pubsub.publish("usersAdded", responseInsertUsers.returning);

      // return the original response
      return responseInsertUsers;
    }
  },
};

module.exports = { resolvers };
```

I get: 
```bash
RangeError: Maximum call stack size exceeded
Exception in PromiseRejectCallback:
C:\dev\github\space-x-land-with-sub\node_modules\@graphql-mesh\runtime\index.cjs.js:142
                const result = await originalResolver(root, args, proxyContext, info);
```

Even if the function is called `originalResolver` in the error message, it's using my new `insert_users` **resolver** and loops until the crash.
Maybe @ardatan could help me here?

Reply:
i know that graphql-tools has an api for wrapping object fields, so assume that is somehow possible https://www.graphql-tools.com/docs/api/classes/_wrap_src_index_.transformobjectfields

Reply:
@jycouet How about using `resolversComposition` instead of overwriting the existing resolvers?
```js
module.exports = next => async (root, args, context, info) => {
  const result = await next(root, args, context, info);
  context.pubsub.publish("usersAdded", responseInsertUsers.returning);
  return result;
}
```

Reply:
Nice way to do @ardatan.
But I find an issue here: what is published highly depends on the original mutation sent.

If I subscribe to
```gql
subscription usersAdded {
  usersAdded {
    id
    name
    timestamp           <- this is not in the insert_user
  }
}
```

And send my mutation:
```gql
insert_users(objects: { name: "JYC", rocket: "codegen" }) {
    affected_rows
    returning {
      id
      name
      rocket
    }
  }
```

I get 
```bash
Cannot return null for non-nullable field users.timestamp.
```

I would like to force field selection for the inside mutation to publish and return the original selection

Reply:
shouldnt the subscription then call the resolver with an query for the selection set?

Reply:
Would be cleaner yes (especially when multiple places can publish later).
But still, for this, I need to ensure that the initial mutation is requesting the id. If I don't request the id, it's not published and I can't retrieve users.

Reply:
Might it be possible to transform the info param AST before forwarding and add the id? 🤔

Reply:
I started looking at this, but I'm not sure it's a good way to go. I could also `throw an error` if you don't request the id... But I'm not a huge fan too.

Here are the different options:
- [x] dedicated mutation to mutate and publish
  - Working well, but not perfect
- [ ] overwriting the mutation at mesh level (I think that this is the best approach)
  - [ ] issue: Infinit loop today
- [ ] composition
  - [ ] updating AST info to pass to `next` ensuring selection of `id`
  - [ ] throw error when user doesn't request the id

I'll speak with the guild to see where to work 👍

Reply:
Damn I just HAD to try it 😂

It works 🤯

```
"use strict";

module.exports = {
    insert_users: next => async (root, args, context, info) => {
        info.operation.selectionSet.selections[0].selectionSet.selections.push({
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "returning"
            },
            "arguments": [],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [
                {
                  "kind": "Field",
                  "name": {
                    "kind": "Name",
                    "value": "id"
                  },
                  "arguments": [],
                  "directives": []
                }
              ]
            }
          })

        const result = await next(root, args, context, info);
        console.log("test", JSON.stringify(result, 0, 2))
        return result;
      }
}
```

Reply:
This would also be a super nice addition for the docs I guess!

Reply:
héhéhé, nice job!!!
It looks like it will be a shared article ;)

You did it following the article? Or from my repo? 

I updated the checklist of options ;)